### ISSUE TITLE
YOLO-NAS: Input and output as frame

### ISSUE THREAD

**Issue by kris-mlguy** [Deci-AI/super-gradients#955]
How to get the input and output from YOLO-NAS in the form of frames where model predictions are with classes and bounding boxes included? I saw a recent update in documentation which indicates access to classes and bounding boxes separately however I'm looking for an option where the output frame is with classes and bounding boxes together and can be saved with cv2.imwrite() method.

**Comment by kris-mlguy** [comment]
Hi @Louis-Dupont,
Thanks for taking it up. I'm looking for 
1) `frame_prediction = model(frame)` ==> frame from video or camera as input
2) `cv2.imwrite(filename, frame_prediction)` ==> output frame (classes and bbox included) can be saved to a file with opencv

**Comment by Louis-Dupont** [comment]
The most efficient way is to first process the whole video, and then draw each frame one by one:

```python
video_predictions = model.predict("<path-to-video")

for frame_index, frame_prediction in enumerate(video_predictions):
    filename = f"frame_{frame_index}.jpg"

    frame_with_bboxes = frame_prediction.draw()
    cv2.imwrite(filename, frame_with_bboxes[:, :, ::-1]) # image is RGB, while cv2 takes BGR

    # Equivalent to:
    # frame_prediction.save(filename)  # save frame as an image
```

You can also do it frame by frame but it will be slower:
```python
for frame_index, frame in enumerate(frames):
    filename = f"frame_{frame_index}.jpg"

    result = model.predict(frame)
    frame_prediction = result[0]  # predict always returns a batch predictions, even when working with single image/frame

    # unchanged
    frame_with_bboxes = frame_prediction.draw()
    cv2.imwrite(filename, frame_with_bboxes[:, :, ::-1]) # image is RGB, while cv2 takes BGR

    # Equivalent to:
    # frame_prediction.save(filename)  # save frame as an image
```

Does that answer your question ?

**Comment by kris-mlguy** [comment]
I wanted to look at frame-wise processing time for comparison and do not want to process the whole video at once. Tried the `frame_prediction.save()` method earlier and it seemed more expensive than `cv2.imwrite()` method on YOLOv8 output frame. Can we have a better method to provide frame_prediction with classes and bbox in real-time that can be used with `cv2.imwrite()` ?

**Comment by Louis-Dupont** [comment]
I would like to better understand some points
- When you say " better method to provide `frame_prediction`": you mean `frame_prediction.save()` that would do the same but faster right?
-  "`frame_prediction.save() ` [...] seemed more expensive": you mean that ` model.predict(frame)` + `frame_prediction.save()` was slow, or did you really mean `frame_prediction.save() ` specifically? Because what takes most time is usually the predict step.
- What exactly are you missing with `frame_with_bboxes = frame_prediction.draw()` in the code below ?

```python
for frame_index, frame in enumerate(frames):
    filename = f"frame_{frame_index}.jpg"

    result = model.predict(frame)
    frame_prediction = result[0]  # predict always returns batch predictions, even when working with a single image/frame

    frame_with_bboxes = frame_prediction.draw()
    cv2.imwrite(filename, frame_with_bboxes[:, :, ::-1]) # image is RGB, while cv2 takes BGR
```

**Comment by kris-mlguy** [comment]
Sure I can explain 

* Either faster `frame_prediction.save()` or as mentioned [here](https://github.com/Deci-AI/super-gradients/issues/955#issuecomment-1540183923)
* `model.predict(frame_jpg)` + `frame_prediction.save()` looks more expensive than YOLOv8 prediction + `cv2.imwrite()`.
* `model.predict()` works with image rather than frame as far as I see in documentation. The feature requested here is essentially the ability to work with single frame efficiently:
`frame_prediction = model.predict(frame)` OR
`frame_prediction = model(frame) ` AND
`frame_with_bboxes = frame_prediction.draw()`
`cv2.imwrite(filename, frame_with_bboxes[:, :, ::-1])`

I'm happy to use if such a feature is already available.

**Comment by Louis-Dupont** [comment]
``` python
import time

import torch
from super_gradients.common.object_names import Models
from super_gradients.training import models
from super_gradients.training.utils.media.video import load_video

# Note that currently only YoloX, PPYoloE and YOLO-NAS are supported.
model = models.get(Models.YOLO_NAS_L, pretrained_weights="coco")

# We want to use cuda if available to speed up inference.
model = model.to("cuda" if torch.cuda.is_available() else "cpu")

frames, _fps = load_video("../../../../documentation/source/images/examples/pose_elephant_flip.gif")

pred_times, save_times = [], []
for i, frame in enumerate(frames):
    start = time.perf_counter()
    image_prediction = model.predict(frame)  # Running on batch = 1, so taking the first prediction.
    end_pred = time.perf_counter()
    pred_times.append(end_pred - start)

    image_prediction.save(f"pose_elephant_flip_{i}.png")
    end_save = time.perf_counter()
    save_times.append(end_save - end_pred)

print("Prediction: ", sum(pred_times))
print("Saving: ", sum(save_times))
print(f"Over: {len(frames)} predictions")
```
```
# on Nvidia RTX A5000
Prediction:  5.186123991268687
Saving:  0.3692859628936276
Over: 77 predictions
```
- The gap would most likely be more important on a smaller GPU or on CPU.
- `save` includes drawing + writing to disk. And this is on GPU

So from this it looks like optimizing the drawing/saving is not what is going to increase dramatically the number of fps processed when running inference in real-time (with `batch_size=1`). Does that make sense, or did I maybe misunderstand your issue ? If I did, it would be great if you can provide a benchmarking script similar to mine so that I can run it myself and be more precise in my answer :)

You can also check out this issue (https://github.com/Deci-AI/super-gradients/issues/959), where we talk about inference speed on pytorch of YoloNAS vs yolov8
